### PR TITLE
[1027][26188] Implement CSP with secure_headers gem

### PR DESCRIPTION
**Pull request by oliverguenther**
**TODOS**

- [x] Remove inline `onclick`, 
- [x] Make link_to_function csp-ready
- [x] Check href="javascript:"
- [x] Rework adminstration user management, uses eval'ed script responses.
  - [x] Removed users/groups JS management
- [x] Check overridden functionality in remote forms of dt customizing (it only decorates the form submission, has nothing to do with remote form handling).
- [x] Rework project settings tabs to avoid js.erb responses (repository)
  - [x] Remove JS tab behavior
- [x] Plugins
  - [x] onclick
  - [x] href="javascript:"

https://community.openproject.com/wp/26188
https://community.openproject.com/wp/1027

### CODE REVIEW

File: app/views/custom_styles/_inline_css.erb [hunk 1]
```diff
@@ -27,7 +27,7 @@ See doc/COPYRIGHT.rdoc for more details.
 
 ++#%>
 
-<style>
+<% nonced_javascript_tag do %>
```
Comment (braiam) [inline app/views/custom_styles/_inline_css.erb]: Shouldn't this be a `hashed_style_tag` instead? [This seems to suggest](https://github.com/twitter/secureheaders/blob/master/docs/hashes.md) that it should be.

File: app/views/my/page_layout.html.erb [hunk 2]
```diff
@@ -27,6 +27,7 @@ See doc/COPYRIGHT.rdoc for more details.
 
 ++#%>
 <script language="JavaScript">
```
Comment (braiam) [inline app/views/my/page_layout.html.erb]: I believe you need to remove the previous script line.

File: app/views/custom_styles/_inline_css.erb (hunk 1 above)
Comment (oliverguenther) [reply app/views/custom_styles/_inline_css.erb]: Not sure what of the article you linked suggests this.

The content is dynamic, thus a statically produced hash (upon asset compilation) would never match. The nonce is generated per-request instead.

File: app/views/my/page_layout.html.erb (hunk 2 above)
Comment (oliverguenther) [reply app/views/my/page_layout.html.erb]: Good catch, thanks!

File: config/initializers/secure_headers.rb [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    # Allow iframe from vimeo (welcome video)
+    frame_src: %w(https://*.vimeo.com 'self'),
+    frame_ancestors: %w('self'),
+    # Allow images from anywhere
+    img_src: %w(* data:),
+    # Allow scripts from self (not inline, but)
+    # for now require unsafe-eval for Angular JIT
+    script_src: %w('self' 'unsafe-eval'),
+    # Allow unsafe-inline styles
+    style_src: assets_src + %w('unsafe-inline'),
```
Comment (wielinde) [inline config/initializers/secure_headers.rb]: @oliverguenther Request for clarification: Isn't the purpose of all this SCP effort to disallow inline styles and JS? Why is it that we trust inline-styles?

Here https://w3c.github.io/webappsec-csp/#csp-directives it says:
> In either case, developers SHOULD NOT include either 'unsafe-inline', or data: as valid sources in their policies. Both enable XSS attacks by allowing code to be included directly in the document itself; they are best avoided completely.

Comment (oliverguenther) [reply config/initializers/secure_headers.rb]: Yeah,  ideally, we can disable inline css as well. But that will kill all `<div style="display: none">..` and other things we have lying around. This is just not feasible at the moment and will blow up the PR by several order of magnitude.

That means that in theory, the cross-site styling attack vector is still open (e.g., a [CSS keylogger](https://css-tricks.com/css-keylogger/) would be possible.

We should try and use `hidden` or semantic classes to set initial styles when rendering from the backend. It's definitely possible, but hard.

The same goes for the `unsafe-eval` used for ng2 JIT. We can disable this as soon as we use the AOT build to compile all templates beforehand.

Reference:
https://stackoverflow.com/a/31759553
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/style-src

Comment (braiam) [reply config/initializers/secure_headers.rb]: @wielinde Sadly, using Angular JIT needs unsave-eval, unlike AoT. https://github.com/angular/angular/issues/12958

Comment (braiam) [reply config/initializers/secure_headers.rb]: @oliverguenther are there plans to switch to AoT?

Comment (oliverguenther) [reply config/initializers/secure_headers.rb]: @braiam Yes, we are going to switch before the angular upgrade goes into production (OpenProject 8.0.). There's been difficulties in bringing AOT and ng1 together so we'll need to see what and when that can happen.

File: app/views/queries/_filters.html.erb [hunk 4]
```diff
@@ -160,8 +160,9 @@ See docs/COPYRIGHT.rdoc for more details.
           <% end %>
         </div>
         <script type="text/javascript">
```
Comment (wielinde) [inline app/views/queries/_filters.html.erb]: I believe this `<script...` line to be obsolete.

File: spec/views/layouts/base.html.erb_spec.rb [hunk 5]
```diff
@@ -198,7 +198,6 @@
         render
         expect(DesignColor.overwritten.size).to eq(1)
         expect(response).to render_template partial: 'custom_styles/_inline_css'
-        expect(rendered).to match /--primary-color:\s*#{primary_color.get_hexcode}/
```
Comment (wielinde) [inline spec/views/layouts/base.html.erb_spec.rb]: If we remove that line there will be no spec left, testing that custom styles really make it into the layout. In other words, this line has some feature spec character.

File: app/assets/javascripts/specific/toggable_fieldset.js [hunk 6]
```diff
@@ -75,14 +75,21 @@ function toggleFieldset(el) {
...
+      var fieldset = getFieldset(this);
+
+      var contentArea = fieldset.find('> div');
+      if (fieldset.hasClass('collapsed')) {
+        contentArea.hide();
+      }
+
+      setFieldsetToggleState(fieldset);
+    });
 });
```
Comment (wielinde) [inline app/assets/javascripts/specific/toggable_fieldset.js:100]: I havn't started debugging, but something is broken here. After collapsing a fieldset it opens straigt away:
bug-fieldset-collapse

File: app/views/common/_tabs.html.erb [hunk 7]
```diff
@@ -40,24 +40,17 @@ See docs/COPYRIGHT.rdoc for more details.
...
-                                       onclick: "showTab('#{tab[:name]}', this.href); this.blur(); return false;") %></li>
+                                       class: "tab-show #{(tab[:name] != selected_tab ? nil : 'selected')}",
+                                       data: { name: tab[:name] }) %></li>
     <% end %>
   </ul>
   <div class="tabs-buttons" style="display:none;">
-    <button class="tab-left icon-context icon-arrow-left4" onclick="moveTabLeft(this);"></button>
-    <button class="tab-right icon-context icon-arrow-right5" onclick="moveTabRight(this);"></button>
+    <button class="tab-left icon-context icon-arrow-left4"></button>
+    <button class="tab-right icon-context icon-arrow-right5"></button>
```
Comment (wielinde) [inline app/views/common/_tabs.html.erb:48]: On my machine the left/right buttons for tabs don't seem to work anymore. When I change back to `dev` branch, they work again.

bug-tab-left-right

File: app/controllers/project_settings_controller.rb [hunk 8]
```diff
@@ -0,0 +1,99 @@
+#-- encoding: UTF-8
```
Comment (wielinde) [inline app/controllers/project_settings_controller.rb:1]: I am not sure where to put this, so I drop it here. When changing the controller name, like you did here, you need to check for implicit i18n consequences. The "no_results_box" is one of those. This text does not get shown properly anymore for instance:         https://github.com/opf/openproject/blob/aa62568debd71a87697ab6f93422a7192091636c/config/locales/en.yml#L143

**Review by wielinde — CHANGES_REQUESTED** [review]
Hey Oliver, here the first findings.

**Comment by oliverguenther** [comment]
@wielinde I believe I have addressed all your remarks. Regarding the renamed no_results, I've added the explicit translations to avoid other languages requiring to update their translation strings.

File: app/assets/javascripts/specific/toggable_fieldset.js (hunk 6 above)
Comment (oliverguenther) [reply app/assets/javascripts/specific/toggable_fieldset.js:100]: Fixed, was caused by the event bubbling up to an upper fieldset. Nested fieldsets.. gwah!

File: app/views/common/_tabs.html.erb (hunk 7 above)
Comment (oliverguenther) [reply app/views/common/_tabs.html.erb:48]: Fixed, was incorrectly assuming a passed event to be a jquery element

**Review by wielinde — APPROVED** [review]

**Comment by wielinde** [comment]
LGTM. Thank you.

**Pull request merged**